---
title: "Reference-guided genome assembly and comparative genomics workflow"
author: "Megan Barkdull"
output: 
  github_document:
    toc: true
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction
The goal of this project is to sequence five new genomes from the ant genus *Cephalotes*, assemble them to the reference *C. varians* genome, and then use these data to better understand the genetic basis of caste polymorphism. 

## Citing
This workflow implements many different tools, all of which should be cited. I have noted appropriate citations throughout the text of this ReadMe. Please properly credit these researchers' hard work!

In addition, if this workflow has been useful to you, please cite the Github repository:

[![DOI](https://zenodo.org/badge/520512524.svg)](https://zenodo.org/badge/latestdoi/520512524)

## Reference-guided genome assembly
Reference-guided genome assembly has the following steps:

  1. Trim raw reads with Trimmomatic.
  2. Map the trimmed reads back to the reference genome (I'll probably use NextGenMap for this).
  3. Call SNPs and indels with SAMtools. 
  4. Filter the calls by quality. 
  5. Assemble a consensus sequence. 
  
These steps are all implemented by the scripts `./Scripts/referenceGuidedAssembly`, which requires two arguments: first, a list of samples, and two, the number of threads to use for multithreaded steps. In other words, run this script with the command `./Scripts/referenceGuidedAssembly [path to file listing samples] [number of threads to use when possible]`.

Note that this script currently has a lot of hard-coded paths, including the paths to each program. The location and names of your raw sequencing files will also be different than those in this script, so update accordingly. 

If you use this script, please cite:

  + Anthony M. Bolger, Marc Lohse, Bjoern Usadel, Trimmomatic: a flexible trimmer for Illumina sequence data, Bioinformatics, Volume 30, Issue 15, 1 August 2014, Pages 2114–2120, https://doi.org/10.1093/bioinformatics/btu170
  + Fritz J. Sedlazeck, Philipp Rescheneder, Arndt von Haeseler, NextGenMap: fast and accurate read mapping in highly polymorphic genomes, Bioinformatics, Volume 29, Issue 21, 1 November 2013, Pages 2790–2791, https://doi.org/10.1093/bioinformatics/btt468
  + Twelve years of SAMtools and BCFtools. Petr Danecek, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, Thomas Keane, Shane A McCarthy, Robert M Davies, Heng Li. GigaScience, Volume 10, Issue 2, February 2021, giab008, https://doi.org/10.1093/gigascience/giab008
  
## Processing assembled genomes
Once genomes are assembled, they require some processing before genes are identified. 

Repetitive regions of the genome must be identified, then soft-masked, prior to annotation. To accomplish this task, we will:

  + Identify repeats in each assembled genome, with RepeatModeler2;
  + Soft mask our identified repeats with RepeatMasker;
  + Soft mask known Hexapod repetitive elements with RepeatMasker, using the RepeatMasker repeat database for Hexapoda. 

All of these steps are implemented in the script `./Scripts/handlingRepeats`, which can be run with the command `./Scripts/handlingRepeats [path to file listing samples] [number of threads to use when possible]`. Again, note that this script uses hard-coded paths to RepeatModeler and RepeatMasker on Cornell's BioHPC, so you'll likely need to change those to reflect your setup. 

Outputs of this step will appear in the directory `./handlingRepeats`.

If you use this script, please cite:

  + Flynn, Jullien M., et al. "RepeatModeler2 for automated genomic discovery of transposable element families." Proceedings of the National Academy of Sciences 117.17 (2020): 9451-9457.
  + Smit, AFA, Hubley, R & Green, P. RepeatMasker Open-4.0.
2013-2015 <http://www.repeatmasker.org>.